File: android-pos/app/src/main/java/com/mcracing/pos/ui/Pricing.kt
package com.mcracing.pos.ui

import java.time.LocalDate

// Mirrors src/lib/pricing.ts so the on-reader "new booking" builder can quote a
// price without a round-trip. Keep in sync with the web matrix.
private val WEEKDAY = mapOf(
    1 to mapOf(1 to 45, 2 to 85, 3 to 115),
    2 to mapOf(1 to 90, 2 to 160, 3 to 220),
    3 to mapOf(1 to 130, 2 to 245, 3 to 340),
)
private val WEEKEND = mapOf(
    1 to mapOf(1 to 50, 2 to 95, 3 to 135),
    2 to mapOf(1 to 100, 2 to 180, 3 to 250),
    3 to mapOf(1 to 140, 2 to 275, 3 to 365),
)

/** Weekend = Fri/Sat/Sun (matches src/lib/pricing.ts isWeekend). */
fun isWeekend(isoDate: String): Boolean {
    return try {
        when (LocalDate.parse(isoDate).dayOfWeek.value) {
            5, 6, 7 -> true // Fri, Sat, Sun
            else -> false
        }
    } catch (_: Exception) {
        false
    }
}

// The venue has 3 sims, so the matrix tops out at 3 racers. A bigger group books
// the same rigs and takes turns, so each racer past the third is a flat add-on
// per hour. MUST match SIM_COUNT / EXTRA_RACER_RATE_PER_HOUR in src/lib/pricing.ts.
const val SIM_COUNT = 3
const val EXTRA_RACER_RATE_PER_HOUR = 10 // dollars

/** The common session lengths, shown as chips. Longer is still bookable. */
val DURATION_OPTIONS = listOf(1.0, 1.5, 2.0, 2.5, 3.0)

/** Longest bookable session — the venue's whole day (noon to 2am close). */
const val MAX_DURATION_HOURS = 14.0

/** Hours past the 3-hour matrix bill per sim seat, per hour. */
const val LONG_SESSION_RATE_PER_SEAT_HOUR = 30

/** "1h" / "1.5h" — chip + summary label. */
fun formatHours(hours: Double): String =
    if (hours % 1.0 == 0.0) "${hours.toInt()}h" else "${hours}h"

/**
 * Session price in cents for (date, any racer count >= 1, 1–3 hours in
 * half-hour steps). The matrix only lists whole hours; a trailing half hour
 * bills at half the one-hour rate. Returns 0 if out of range.
 */
fun sessionPriceCents(isoDate: String, racers: Int, hours: Double): Long {
    val matrix = if (isWeekend(isoDate)) WEEKEND else WEEKDAY
    val seated = racers.coerceIn(1, SIM_COUNT)
    val withinMatrix = hours.coerceIn(1.0, 3.0)
    val whole = Math.floor(withinMatrix).toInt()
    val hasHalf = withinMatrix - whole >= 0.5
    val wholePrice = matrix[seated]?.get(whole) ?: return 0L
    val oneHourPrice = matrix[seated]?.get(1) ?: return 0L
    // A trailing half hour bills at HALF THE ONE-HOUR RATE — it must NOT
    // interpolate toward the next tier, which would hand out part of the 2h
    // bulk discount. Keep in sync with matrixPrice() in src/lib/pricing.ts.
    val base = wholePrice + (if (hasHalf) oneHourPrice / 2.0 else 0.0)
    // Past 3 hours the matrix runs out — bill the flat per-seat rate per hour.
    val longHours = (hours - 3.0).coerceAtLeast(0.0)
    val longCharge = longHours * LONG_SESSION_RATE_PER_SEAT_HOUR * seated
    val extras = (racers - SIM_COUNT).coerceAtLeast(0)
    val dollars = base + longCharge + extras * EXTRA_RACER_RATE_PER_HOUR * hours
    return Math.round(dollars * 100.0)
}

// Sales tax — MUST match src/lib/tax.ts (SALES_TAX_RATE_BPS) + the backend env.
// The backend is authoritative for what's charged; this only drives the on-reader
// display so the customer sees the same total they'll be charged. If the rate
// changes, update the backend env AND this constant, then rebuild the app.
const val SALES_TAX_RATE_BPS = 700 // 7.00%

/** Tax on a pre-tax subtotal, in cents (rounded to nearest cent, matches backend). */
fun computeTaxCents(subtotalCents: Long): Long {
    if (subtotalCents <= 0) return 0L
    return Math.round(subtotalCents.toDouble() * SALES_TAX_RATE_BPS / 10000.0)
}

/** e.g. "7%" — for the reader's tax line. */
fun taxRateLabel(): String {
    val pct = SALES_TAX_RATE_BPS / 100.0
    return if (pct == Math.floor(pct)) "${pct.toLong()}%" else "%.2f%%".format(pct)
}

File: src/lib/pricing.ts
// Pricing configuration for MC Racing Sim
// Weekday: Tuesday-Thursday
// Weekend: Friday-Sunday
// Monday: Closed

/** Hours on the matrix. Half-hours in between are priced by interpolation. */
type MatrixHours = 1 | 2 | 3
/** Session length in hours — any half-hour step from 1 to 3. */
export type Duration = number
/** Racers on the matrix — one per sim. Beyond this, see EXTRA_RACER_RATE. */
type SeatedRacers = 1 | 2 | 3

/** The common session lengths, shown as tiles/chips. Longer is still bookable. */
export const DURATION_OPTIONS: readonly number[] = [1, 1.5, 2, 2.5, 3]

/** Longest bookable session — the venue's whole day (noon to 2am close). */
export const MAX_DURATION_HOURS = 14

/**
 * Hours past the 3-hour matrix are billed per SIM SEAT, per hour. Racers beyond
 * the sims keep their cheaper EXTRA_RACER_RATE_PER_HOUR for the whole session —
 * they're sharing a rig, not occupying a fourth one.
 */
export const LONG_SESSION_RATE_PER_SEAT_HOUR = 30

/** Every bookable length in half-hour steps, for admin dropdowns. */
export function allDurationOptions(): number[] {
  const out: number[] = []
  for (let h = 1; h <= MAX_DURATION_HOURS; h += 0.5) out.push(h)
  return out
}

/** A bookable length: 1 hour up to a full day, on a half-hour step. */
export function isValidDuration(hours: number): boolean {
  return (
    Number.isFinite(hours) &&
    hours >= 1 &&
    hours <= MAX_DURATION_HOURS &&
    Number.isInteger(hours * 2) // half-hour steps only
  )
}

/** "1 hour" / "1.5 hours" / "2 hours" — one label rule for every surface. */
export function formatDuration(hours: number): string {
  const n = Number(hours)
  const text = Number.isInteger(n) ? String(n) : n.toFixed(1).replace(/\.0$/, '')
  return `${text} hour${n === 1 ? '' : 's'}`
}
/** Any racer count on a booking. 1–3 fill the sims; extras rotate through them. */
export type RacerCount = number

/**
 * The venue has 3 sims, so the price matrix tops out at 3 racers. A bigger group
 * still books the same 3 rigs and takes turns, so each racer beyond the third is
 * a flat add-on per hour rather than a full seat price.
 *
 * Must match SIM_COUNT / EXTRA_RACER_RATE in android-pos/.../ui/Pricing.kt.
 */
export const SIM_COUNT = 3
export const EXTRA_RACER_RATE_PER_HOUR = 10 // dollars, per racer past the 3rd

/**
 * Group size that gets a "did you mean that?" confirm before it's booked.
 * There's no hard cap, so this is the guard against a fat-fingered 50 — which
 * would quietly hold the whole venue and quote a four-figure session.
 */
export const LARGE_GROUP_RACERS = 10

const WEEKDAY_PRICES: Record<SeatedRacers, Record<MatrixHours, number>> = {
  1: { 1: 45, 2: 85, 3: 115 },
  2: { 1: 90, 2: 160, 3: 220 },
  3: { 1: 130, 2: 245, 3: 340 },
}

const WEEKEND_PRICES: Record<SeatedRacers, Record<MatrixHours, number>> = {
  1: { 1: 50, 2: 95, 3: 135 },
  2: { 1: 100, 2: 180, 3: 250 },
  3: { 1: 140, 2: 275, 3: 365 },
}

/** Racers charged the flat hourly add-on (everyone past the 3rd). */
export function extraRacers(racerCount: RacerCount): number {
  return Math.max(0, Math.floor(racerCount) - SIM_COUNT)
}

/** Dollar add-on for the extra racers over the whole session. */
export function extraRacerChargeDollars(
  racerCount: RacerCount,
  duration: Duration
): number {
  return extraRacers(racerCount) * EXTRA_RACER_RATE_PER_HOUR * duration
}

export function isWeekend(date: Date | string): boolean {
  const d = typeof date === 'string' ? new Date(date + 'T12:00:00') : date
  const day = d.getDay()
  // Friday = 5, Saturday = 6, Sunday = 0
  return day === 0 || day === 5 || day === 6
}

export function isMonday(date: Date | string): boolean {
  const d = typeof date === 'string' ? new Date(date + 'T12:00:00') : date
  return d.getDay() === 1
}

/**
 * Matrix price for a seated racer count at any half-hour length. The matrix
 * only lists whole hours, so a half-hour lands exactly between its neighbours
 * (e.g. 1.5h = midway between the 1h and 2h price).
 */
function matrixPrice(
  matrix: Record<SeatedRacers, Record<MatrixHours, number>>,
  seated: SeatedRacers,
  hours: number
): number {
  const clamped = Math.min(Math.max(hours, 1), 3)
  const whole = Math.floor(clamped) as MatrixHours
  const hasHalf = clamped - whole >= 0.5

  // A trailing half hour bills at HALF THE ONE-HOUR RATE, on top of the whole
  // hours. It deliberately does NOT interpolate toward the next tier: the 2h
  // and 3h prices carry a bulk discount, and interpolating handed a 1.5h
  // session part of the 2-hour discount it hasn't earned.
  //   weekend 3 racers: 1h $140 -> 1.5h = 140 + 70 = $210  (interpolation gave $207.50)
  const within = matrix[seated][whole] + (hasHalf ? matrix[seated][1] / 2 : 0)

  // Past 3 hours the matrix runs out, so every further hour bills at the flat
  // per-seat long-session rate.
  const longHours = Math.max(0, hours - 3)
  return within + longHours * LONG_SESSION_RATE_PER_SEAT_HOUR * seated
}

export function calculatePrice(
  date: Date | string,
  duration: Duration,
  racerCount: RacerCount
): { price: number; isWeekend: boolean } {
  const weekend = isWeekend(date)
  const priceMatrix = weekend ? WEEKEND_PRICES : WEEKDAY_PRICES

  // Up to 3 racers price straight off the matrix. Past that the group is still
  // on the same 3 sims (taking turns), so each additional racer is the flat
  // hourly add-on on top of the full 3-racer rate.
  const seated = Math.min(Math.max(Math.floor(racerCount), 1), SIM_COUNT) as SeatedRacers
  const price =
    matrixPrice(priceMatrix, seated, duration) +
    extraRacerChargeDollars(racerCount, duration)

  // Half-hours can land on a half-dollar (e.g. $187.50); round to the cent so
  // no float dust reaches the charge.
  return { price: Math.round(price * 100) / 100, isWeekend: weekend }
}

export function getDayType(date: Date | string): 'weekday' | 'weekend' | 'closed' {
  if (isMonday(date)) return 'closed'
  return isWeekend(date) ? 'weekend' : 'weekday'
}

export function formatPrice(price: number): string {
  return `$${price}`
}

// Get all time slots for operating hours (noon to 2am)
export function getTimeSlots(): string[] {
  const slots: string[] = []

  // From noon (12:00) to 11:30 PM
  for (let hour = 12; hour <= 23; hour++) {
    const period = hour >= 12 ? 'PM' : 'AM'
    const displayHour = hour % 12 || 12
    slots.push(`${displayHour}:00 ${period}`)
    slots.push(`${displayHour}:30 ${period}`)
  }

  // From 12:00 AM to 1:30 AM (closing is 2am)
  slots.push('12:00 AM')
  slots.push('12:30 AM')
  slots.push('1:00 AM')
  slots.push('1:30 AM')

  return slots
}

export function formatTime(time: string): string {
  return time
}

export function formatDate(dateStr: string): string {
  const date = new Date(dateStr + 'T12:00:00')
  return date.toLocaleDateString('en-US', {
    weekday: 'short',
    month: 'short',
    day: 'numeric',
  })
}

export function formatDateLong(dateStr: string): string {
  const date = new Date(dateStr + 'T12:00:00')
  return date.toLocaleDateString('en-US', {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  })
}

// ============================================================================
// No-show fee
// ============================================================================
// Per spec (June 23, 2026): flat $20 per seat booked, charged to card on file
// if the customer no-shows. Stored on the booking at creation time so we can't
// retroactively change what they consented to.

export const NO_SHOW_FEE_CENTS_PER_SEAT = 2000 // $20.00

/**
 * Charged per SIM SEAT held, not per racer — a no-show only ever costs the
 * venue the 3 rigs, so a 5-racer group's fee caps at 3 × $20 = $60.
 */
export function calculateNoShowFeeCents(racerCount: RacerCount): number {
  const seatsHeld = Math.min(Math.max(Math.floor(racerCount), 1), SIM_COUNT)
  return seatsHeld * NO_SHOW_FEE_CENTS_PER_SEAT
}

export function formatNoShowFee(racerCount: RacerCount): string {
  const cents = calculateNoShowFeeCents(racerCount)
  return `$${(cents / 100).toFixed(0)}`
}
